src/history/filter.rs

```diff
@@ -0,0 +1,167 @@
+use crate::{
+    result::{ReedlineError, ReedlineErrorVariants},
+    History, HistoryItem, HistoryItemId, Result, SearchDirection, SearchFilter, SearchQuery,
+};
+
+/// A history that wraps another history, and does not forward values beginning with `exclusion_prefix` (if present).
+/// If an item is filtered, it is stored in memory and able to be retrieved, until the next item is inserted.
+#[derive(Debug)]
+pub struct HistoryFilter<T> {
```

Comment:
Based on the usage pattern in `src/engine.rs` I would make this `pub(crate)` only. To me, through the builder pattern it looks like an internal implementation detail.